Explore to SQL Lab

betodealmeida opened the pull request:
This PR adds a button that takes from the explore view to SQLlab, with a pre-populated query on the datasource:

![sqllab](https://user-images.githubusercontent.com/1534870/40692457-c8be5bfa-6366-11e8-8060-f33606ff25ff.gif)

betodealmeida committed:
WIP

betodealmeida committed:
Working version

betodealmeida committed:
Clean code

betodealmeida committed:
Fix lint

betodealmeida committed:
Fix unit test; show only for sqla

mistercrunch commented:
The link should only show for SQLAlchemy datasources. I've been meaning to do something similar from the "View query" modal as well.

betodealmeida commented:
@mistercrunch it only shows for SQLA: https://github.com/apache/incubator-superset/pull/5101/files#diff-39c3f576c74efa400932ceb7eebe8a91R214

I talked with @vylc and she thinks it's more useful to pre-generate the query based on the form data, instead of a plain `SELECT *`. I'll update this PR.

mistercrunch commented:
Yup there was some thoughts of showing a "Preview data" tab here as well.

betodealmeida committed:
Working on UX

betodealmeida committed:
Dropdown working 66%

betodealmeida committed:
Working but needs CSS

betodealmeida committed:
Fixed table

betodealmeida commented:
I changed the "View Query" button to a dropdown label "Query" with 3 options:

- View query: same as before
- View data: shows preview of the data
- Run in SQL Lab: auto-runs the query in SQL Lab

For the data preview, we can't reuse the data that is present in `queryResponse`, since its format depends on the viz type. Instead, I changed the `query` endpoint to return the data in a format appropriate for displaying as a table.


![explore_to_sqllab](https://user-images.githubusercontent.com/1534870/40867380-a70ae2d6-65b8-11e8-9666-9b0b91817ab3.gif)

betodealmeida commented on superset/assets/src/chart/chartAction.js:
```diff
@@ -186,3 +186,23 @@ export function runQuery(formData, force = false, timeout = 60, key) {
+        }
+        url.searchParams.set('datasourceKey', formData.datasource);
+        url.searchParams.set('sql', response.query);
+        window.open(url.href, '_blank');
```
@mistercrunch, is there a better way of doing this? I wasn't sure how to get the base URL from JS.

mistercrunch replied:
I think it's a bit the wild west in terms of URL composition in Superset at the moment. It could be good to refactor a `src/utils/urls.js` module with constants and utilities for all this.

betodealmeida committed:
Fix lint

betodealmeida committed:
Fix unit test

mistercrunch commented on superset/assets/src/SqlLab/actions.js:
```diff
@@ -405,6 +405,25 @@ export function popSavedQuery(saveQueryId) {
+        };
+        dispatch(addQueryEditor(queryEditorProps));
+      },
+      error: () => notify.error(t('The datasource couldn\'t be loaded')),
```
nit (optional): the linter allows double quotes when the string contains a single quote, so that you don't have to escape it

mistercrunch commented on superset/assets/src/explore/components/DisplayQueryButton.jsx:
```diff
@@ -6,13 +6,17 @@ import markdown from 'react-syntax-highlighter/dist/languages/markdown';
 import Button from '../../components/Button';
 import { t } from '../../locales';
 
+require('react-bootstrap-table/css/react-bootstrap-table.css');
```
nit: import should do the same

mistercrunch commented on superset/assets/src/explore/components/DisplayQueryButton.jsx:
```diff
@@ -21,6 +25,7 @@ registerLanguage('json', json);
 const $ = window.$ = require('jquery');
 
 const propTypes = {
+  actions: PropTypes.object.isRequired,
```
(optional) I think the code has a fair amount of passing the whole `actions` object around. I think it's preferable to pass something more generic when possible to make the component more reusable and less bound to the app. Here you could pass the `redirectSQLLab` action to an `onOpenInEditor` prop.

betodealmeida replied:
Good point, will do that.

mistercrunch commented on superset/assets/src/explore/components/controls/DatasourceControl.jsx:
```diff
@@ -211,6 +211,18 @@ export default class DatasourceControl extends React.PureComponent {
             />
           </a>
         </OverlayTrigger>
+        {this.props.datasource.type === 'table' && <OverlayTrigger
```
nit (optional): personally I like  to break before an incomplete tag as in `\n<OverlayTrigger` but will defer on the linter here

mistercrunch commented on superset/connectors/base/models.py:
```diff
@@ -152,6 +152,12 @@ def short_data(self):
             'creator': str(self.created_by),
         }
 
+    @property
```
This method belongs in `connectors.sqla.models.SqlaTable` as Druid datasources don't linke to `Database` but to cluster, this base should have a `raise NotImplementedError()` or `pass`.

Also in that context you'll be able to simply use `self.database.select_star(...)`

mistercrunch commented on superset/connectors/base/models.py:
```diff
@@ -185,6 +191,8 @@ def data(self):
             'columns': [o.data for o in self.columns],
             'verbose_map': verbose_map,
+            'schema': self.schema,
+            'select_star': self.select_star,
```
Given this here in the base class I'd go for a `pass`  or `return None` in `BaseDatasource.select_star`

mistercrunch commented on superset/views/core.py:
```diff
@@ -1038,6 +1038,18 @@ def get_viz(
 
+    @has_access
+    @expose('/sql')
+    def sql_from_form_data(self):
```
I feel like this endpoint already exists, let me check.

mistercrunch replied:
Basically call `getExploreUrlAndPayload(formData, endpointType = true)`

mistercrunch replied:
FYI it gets called when opening the "View Query" while the query is still running or has never run yet.

mistercrunch commented:
A few minor comments, otherwise LGTM

betodealmeida committed:
Fix merge conflict

betodealmeida committed:
Fix languages path

betodealmeida committed:
Fixes

betodealmeida committed:
Fix Javascript lint
